src/v/datalake/values_avro.cc

```diff
@@ -0,0 +1,278 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "datalake/values_avro.h"
+
+#include "bytes/iobuf_parser.h"
+#include "serde/avro/parser.h"
+
+namespace datalake {
+
+namespace {
+
+ss::future<value_outcome> avro_parsed_to_value(
+  const std::unique_ptr<serde::avro::parsed::message>& parsed_msg,
+  avro::NodePtr node);
+
+template<typename T, typename ValueT>
+value_outcome
+convert_primitive(T v, avro::Type expected_type, const avro::NodePtr& node) {
+    if (node->type() != expected_type) {
+        return value_conversion_exception(fmt::format(
+          "Schema mismatch detected expected type {} got {}, current node: {}",
+          expected_type,
+          node->type(),
+          node));
+    }
+    return ValueT{std::move(v)};
+}
+
+struct primitive_visitor {
+    using ret_t = value_outcome;
+    ret_t operator()(int32_t v) {
+        if (node->logicalType().type() == avro::LogicalType::DATE) {
+            return convert_primitive<int32_t, iceberg::date_value>(
+              v, avro::AVRO_INT, node);
+        }
+        if (node->logicalType().type() == avro::LogicalType::TIME_MILLIS) {
+            std::chrono::milliseconds ms(v);
+            return convert_primitive<int64_t, iceberg::time_value>(
+              std::chrono::duration_cast<std::chrono::microseconds>(ms).count(),
+              avro::AVRO_INT,
+              node);
+        }
+        return convert_primitive<int32_t, iceberg::int_value>(
+          v, avro::AVRO_INT, node);
+    }
+    ret_t operator()(int64_t v) {
+        if (node->type() == avro::AVRO_ENUM) {
+            return convert_primitive<int64_t, iceberg::long_value>(
+              v, avro::AVRO_ENUM, node);
+        }
+        if (node->logicalType().type() == avro::LogicalType::TIME_MICROS) {
+            return convert_primitive<int64_t, iceberg::time_value>(
+              v, avro::AVRO_LONG, node);
+        }
+        if (node->logicalType().type() == avro::LogicalType::TIME_MICROS) {
+            return convert_primitive<int64_t, iceberg::time_value>(
+              v, avro::AVRO_LONG, node);
```

Comment:
Meant to be TIME_MILLIS? Otherwise this is a dupe of the above branch

Reply:
it is redundant, time in milliseconds is an int